In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# **Challenge_TelecomX_LATAM - grupo G9**
**Desarrollado : Jaime Pradenas **
** Diciembre 2025/Enero 2026 **

- Pantilla Trello - https://trello.com/b/KlMd7hug/telecomxlatamjpraden
- Repositorio Github - https://github.com/jpraden/Challenge-DataScience-TelecomX-LATAM/tree/main

💡**Acerca del desafío**💡

**Descripción**

Telecom X - Análisis de Evasión de Clientes
Has sido contratado como asistente de análisis de datos en Telecom X y formarás parte del proyecto "Churn de Clientes". La empresa enfrenta una alta tasa de cancelaciones y necesita comprender los factores que llevan a la pérdida de clientes.

Tu desafío será recopilar, procesar y analizar los datos, utilizando Python y sus principales bibliotecas para extraer información valiosa. A partir de tu análisis, el equipo de Data Science podrá avanzar en modelos predictivos y desarrollar estrategias para reducir la evasión.

**¿Qué vas a practicar?**

✅ Importar y manipular datos desde una API de manera eficiente.

✅ Aplicar los conceptos de ETL (Extracción, Transformación y Carga) en la preparación de los datos.

✅ Crear visualizaciones estratégicas para identificar patrones y tendencias.

✅ Realizar un Análisis Exploratorio de Datos (EDA) y generar un informe con insights relevantes.


¡Ahora es tu turno! 🚀 Usa tus conocimientos para transformar datos en información estratégica y ayudar a Telecom X a retener más clientes.




#📌 Extracción

**Descripción:**


El principal objetivo de este desafío es desarrollar tus habilidades en ETL en este caso (Extract) con Python. Los datos para este desafío están disponibles en una API. Estos datos están disponibles en formato JSON y contienen información esencial sobre los clientes, incluyendo datos demográficos, tipo de servicio contratado y estado de evasión.

**Acceso a los datos para extraer en tú local:**

🔗 Telecom X -> https://github.com/alura-cursos/challenge2-data-science-LATAM/blob/main/TelecomX_Data.json

**Acceso a los datos desde API para usar  en tú código:**

🔗 Telecom X --> https://raw.githubusercontent.com/ingridcristh/challenge2-data-science/refs/heads/main/TelecomX_Data.json

______________________________________________________________________________


**Actividades a realizar en esta primera etapa:**

✅ Importar  desde una API de manera eficiente.

✅ Comprender qué información contiene el dataset (Estructura y datos).



📚 **Bibliotecas y librerias para extracción, exploración y manipulación de datos**
 - import pandas as pd
 - import numpy as np
 - import requests
 - import json


In [2]:
#*Bibliotecas y librerias para extracción, exploración y manipulación de datos**
import pandas as pd
import numpy as np
import requests
import json


In [3]:
#Obteniendo datos desde API (Extracción desde repositorio de github en  )
#url = 'https://raw.githubusercontent.com/jpraden/Challenge-DataScience-TelecomX-LATAM/main/TelecomX_Data.json'
#urlAPI = 'https://raw.githubusercontent.com/alura-cursos/challenge2-data-science-LATAM/main/TelecomX_Data.json'
urlAPI = 'https://raw.githubusercontent.com/ingridcristh/challenge2-data-science/refs/heads/main/TelecomX_Data.json'

response = requests.get(urlAPI)
data = response.json()
df = pd.DataFrame(data)
df.head()

customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   customerID  7267 non-null   object
 1   Churn       7267 non-null   object
 2   customer    7267 non-null   object
 3   phone       7267 non-null   object
 4   internet    7267 non-null   object
 5   account     7267 non-null   object
dtypes: object(6)
memory usage: 340.8+ KB


In [5]:
df.shape

(7267, 6)

**🔍** **Información de la Extracción de datos desde ORIGEN**

  - **filas** - `'7.267'`: Cantidad de registros.
  - **columnas** - `'6'`: Atributos de los datos.

Los datos están organizados en un formato jerárquico (diccionario dentro de diccionario), conteniendo la siguiente información:

**🔍** **Análisis de la Estructura de los datos**

1. **Identificación del Cliente** - `'customerID'`: Identificador único del cliente.
* `'Churn'`: Indica si el cliente canceló el servicio (`Yes` o `No`).

2. **Información del Cliente** (`'customer'`)
* `'gender'`: Género del cliente (`Male` o `Female`).
* `'SeniorCitizen'`: Indica si el cliente es una persona mayor (0 = No, 1 = Sí).
* `'Partner'`: Si el cliente tiene pareja.
* `'Dependents'`: Si el cliente tiene dependientes.
* `'tenure'`: Tiempo de permanencia como cliente (en meses).


3. **Servicios de Telefonía** (`'phone'`)
* `'PhoneService'`: Indica si el cliente posee servicio de teléfono (`Yes` o `No`).
* `'MultipleLines'`: Si posee múltiples líneas telefónicas.


4. **Servicios de Internet** (`'internet'`)
* Tipo de servicio de internet contratado (`DSL`, `Fiber optic`, `No`).
* Servicios adicionales (`OnlineSecurity`, `OnlineBackup`, `DeviceProtection`, `TechSupport`, `StreamingTV`, `StreamingMovies`).


5. **Información de la Cuenta** (`'account'`)
* `'Contract'`: Tipo de contrato (`Month-to-month`, `One year`, `Two year`).
* `'PaperlessBilling'`: Si el cliente recibe facturas electrónicas.
* `'PaymentMethod'`: Método de pago (`Bank transfer`, `Credit card`, `Electronic check`, `Mailed check`).
* `'Charges'`:
* `'Monthly'`: Valor mensual cobrado.
* `'Total'`: Valor total pagado por el cliente.

#🔧 Transformación
**Descripción:**

El principal objetivo de este desafío es desarrollar tus habilidades en ETL en este caso (Transform) con Python. Los datos para este desafío están disponibles en una API.

En este desafío, el enfoque está en el proceso de extracción de datos desde la API, limpieza y transformación.

**Actividades a realizar en esta primera etapa:**


✅ 1. Normalizar el dataset.

✅ 2. Verificar valores ùnicos, nulos, vacíos del dataset normalizado.

✅ 3. Tranformar y corregir las inconsistencias.

✅ 4. Crear la columna de cuentas diarias (Opcional).

✅ 5. Traducir las columnas y/o datos.



# 1. Normalizamos la dataset

In [6]:
# Normalizamos la dataset
df = pd.json_normalize(data)
df

customerID Churn customer.gender  customer.SeniorCitizen  \
0     0002-ORFBO    No          Female                       0   
1     0003-MKNFE    No            Male                       0   
2     0004-TLHLJ   Yes            Male                       0   
3     0011-IGKFF   Yes            Male                       1   
4     0013-EXCHZ   Yes          Female                       1   
...          ...   ...             ...                     ...   
7262  9987-LUTYD    No          Female                       0   
7263  9992-RRAMN   Yes            Male                       0   
7264  9992-UJOEL    No            Male                       0   
7265  9993-LHIEB    No            Male                       0   
7266  9995-HOTOH    No            Male                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
0                 Yes                 Yes                9                Yes   
1                  No                  No                9                Yes   
2                  No                  No                4                Yes   
3                 Yes                  No               13                Yes   
4                 Yes                  No                3                Yes   
...               ...                 ...              ...                ...   
7262               No                  No               13                Yes   
7263              Yes                  No               22                Yes   
7264               No                  No                2                Yes   
7265              Yes                 Yes               67                Yes   
7266              Yes                 Yes               63                 No   

     phone.MultipleLines internet.InternetService  ... internet.OnlineBackup  \
0                     No                      DSL  ...                   Yes   
1                    Yes                      DSL  ...                    No   
2                     No              Fiber optic  ...                    No   
3                     No              Fiber optic  ...                   Yes   
4                     No              Fiber optic  ...                    No   
...                  ...                      ...  ...                   ...   
7262                  No                      DSL  ...                    No   
7263                 Yes              Fiber optic  ...                    No   
7264                  No                      DSL  ...                   Yes   
7265                  No                      DSL  ...                    No   
7266    No phone service                      DSL  ...                   Yes   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
0                           No                  Yes                  Yes   
1                           No                   No                   No   
2                          Yes                   No                   No   
3                          Yes                   No                  Yes   
4                           No                  Yes                  Yes   
...                        ...                  ...                  ...   
7262                        No                  Yes                   No   
7263                        No                   No                   No   
7264                        No                   No                   No   
7265                       Yes                  Yes                   No   
7266                       Yes                   No                  Yes   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
0                          No         One year                      Yes   
1                         Yes   Month-to-month                       No   
2                          No   Month-to-month                      Yes   
3                         Yes   Month-to-month                    

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7267 non-null   object 
 1   Churn                      7267 non-null   object 
 2   customer.gender            7267 non-null   object 
 3   customer.SeniorCitizen     7267 non-null   int64  
 4   customer.Partner           7267 non-null   object 
 5   customer.Dependents        7267 non-null   object 
 6   customer.tenure            7267 non-null   int64  
 7   phone.PhoneService         7267 non-null   object 
 8   phone.MultipleLines        7267 non-null   object 
 9   internet.InternetService   7267 non-null   object 
 10  internet.OnlineSecurity    7267 non-null   object 
 11  internet.OnlineBackup      7267 non-null   object 
 12  internet.DeviceProtection  7267 non-null   object 
 13  internet.TechSupport       7267 non-null   objec

In [8]:
df.shape

(7267, 21)

#2. Verificar valores

1. Verificando valores unicos

In [9]:
for col in df.columns:
  print(f'{col}: {df[col].nunique()}')
  if df[col].nunique() < 50:
    print(df[col].unique())
    print('-' * 50)

customerID: 7267
Churn: 3
['No' 'Yes' '']
--------------------------------------------------
customer.gender: 2
['Female' 'Male']
--------------------------------------------------
customer.SeniorCitizen: 2
[0 1]
--------------------------------------------------
customer.Partner: 2
['Yes' 'No']
--------------------------------------------------
customer.Dependents: 2
['Yes' 'No']
--------------------------------------------------
customer.tenure: 73
phone.PhoneService: 2
['Yes' 'No']
--------------------------------------------------
phone.MultipleLines: 3
['No' 'Yes' 'No phone service']
--------------------------------------------------
internet.InternetService: 3
['DSL' 'Fiber optic' 'No']
--------------------------------------------------
internet.OnlineSecurity: 3
['No' 'Yes' 'No internet service']
--------------------------------------------------
internet.OnlineBackup: 3
['Yes' 'No' 'No internet service']
--------------------------------------------------
internet.DeviceProtecti

2. Verificando valores duplicados

In [10]:
#Verificando valores duplicados
print("Numero de duplicados", df.duplicated().sum())

Numero de duplicados 0


3. Verificando valores nulos

In [11]:
#Verificando valores nulos
print("Numero de Nulos \n", df.isnull().sum())

Numero de Nulos 
 customerID                   0
Churn                        0
customer.gender              0
customer.SeniorCitizen       0
customer.Partner             0
customer.Dependents          0
customer.tenure              0
phone.PhoneService           0
phone.MultipleLines          0
internet.InternetService     0
internet.OnlineSecurity      0
internet.OnlineBackup        0
internet.DeviceProtection    0
internet.TechSupport         0
internet.StreamingTV         0
internet.StreamingMovies     0
account.Contract             0
account.PaperlessBilling     0
account.PaymentMethod        0
account.Charges.Monthly      0
account.Charges.Total        0
dtype: int64


4. Verificando valores vacios o blancos

In [12]:
#Verificando valores vacios o blancos
df.apply(lambda x: x.astype(str).str.strip() == '').sum()

customerID                     0
Churn                        224
customer.gender                0
customer.SeniorCitizen         0
customer.Partner               0
customer.Dependents            0
customer.tenure                0
phone.PhoneService             0
phone.MultipleLines            0
internet.InternetService       0
internet.OnlineSecurity        0
internet.OnlineBackup          0
internet.DeviceProtection      0
internet.TechSupport           0
internet.StreamingTV           0
internet.StreamingMovies       0
account.Contract               0
account.PaperlessBilling       0
account.PaymentMethod          0
account.Charges.Monthly        0
account.Charges.Total         11
dtype: int64

**🔍** **Resultados de normalización del dataset**

  - **filas** - `'7.267'`: Cantidad de registros.
  - **columnas** - `'21'`: Atributos de los datos.

___________________________________________________________________________  

___________________________________________________________________________

**1. Valores Únicos -> Descripción atributos nuevo Dataset Normalizado**

1. `'customerID'`: Identificador único del cliente.
2. `'Churn'`: Indica si el cliente canceló el servicio (`Yes` o `No` o '').
3. `'customer.gender'`: Género del cliente (`Male` o `Female`).
4. `'customer.SeniorCitizen'`: Indica si el cliente es una persona mayor (0 = No, 1 = Sí).
5. `'customer.Partner'`: Si el cliente tiene pareja (`Yes` o `No`).
6. `'customer.Dependents'`: Si el cliente tiene dependientes (`Yes` o `No`).
7. `'customer.tenure'`: Tiempo de permanencia como cliente (en meses).
8. `'phone.PhoneService'`: Indica si el cliente posee servicio de teléfono (`Yes` o `No`).
9. `'phone.MultipleLines'`: Si posee múltiples líneas telefónicas. (`Yes` o `No`, `No phone service`).
10. `'internet.InternetService'`: Servicio de internet (`DSL`, `Fiber optic`, `No`).
11. `'internet.OnlineSecurity'`: Servicio internet de seguridad (`No`, `Yes`, `No internet service`).
12. `'internet.OnlineBackup'`: Servicio internet de respaldo (`No`, `Yes`, `No internet service`).
13. `'internet.DeviceProtection'`: Servicio internet proteciòn (`No`, `Yes`, `No internet service`).
14. `'internet.TechSupport'`: Servicio internet soporte tècnico (`No`, `Yes`, `No internet service`).
15. `'internet.StreamingTV'`: Servicio internet streaming de televisión (`No`, `Yes`, `No internet service`).
16. `'internet.StreamingMovies'`: Servicio internet streaming de peliculas (`No`, `Yes`, `No internet service`).
17. `'account.Contract'`: Tipo de contrato (`Month-to-month`, `One year`, `Two year`).
18. `'account.PaperlessBilling'`: Si el cliente recibe facturas electrónicas (`Yes` o `No`).
19. `'account.PaymentMethod'`: Método de pago (`Bank transfer`, `Credit card`, `Electronic check`, `Mailed check`).
20. `'account.Charges.Monthly'`: Valor mensual cobrado.
21. `'account.Charges.Total'`: Valor total pagado por el cliente.

______________________________________________________________________________


**2. No se registran valores duplicados en el dataset normalizado**

______________________________________________________________________________


**3. No se registran valores nulos en el dataset normalizado**

______________________________________________________________________________


**4. Se registran valores vacíos o blancos**, en atributos:
- Churn	: 224 registros.
- account.Charges.Total :	11 registros.

______________________________________________________________________________



#3. Tranformar y corregir las inconsistencias.

1. Cambiando la columna "account.Charges.Total" a Float y columna "Churn"

In [13]:
#1. Cambiando la columna "account.Charges.Total" a Float
df['account.Charges.Total'] = pd.to_numeric(df['account.Charges.Total'], errors='coerce')

In [14]:
df= df[df['Churn'].str.strip() != '']
print("Número de filas después de eliminar las vacías en 'Churn':", len(df))


Número de filas después de eliminar las vacías en 'Churn': 7043


2. Verificación de atributos nulos, vacíos o en blanco

In [15]:
df.apply(lambda x: x.astype(str).str.strip() == '').sum()

customerID                   0
Churn                        0
customer.gender              0
customer.SeniorCitizen       0
customer.Partner             0
customer.Dependents          0
customer.tenure              0
phone.PhoneService           0
phone.MultipleLines          0
internet.InternetService     0
internet.OnlineSecurity      0
internet.OnlineBackup        0
internet.DeviceProtection    0
internet.TechSupport         0
internet.StreamingTV         0
internet.StreamingMovies     0
account.Contract             0
account.PaperlessBilling     0
account.PaymentMethod        0
account.Charges.Monthly      0
account.Charges.Total        0
dtype: int64

In [16]:
print("Numero de Nulos \n", df.isnull().sum())

Numero de Nulos 
 customerID                    0
Churn                         0
customer.gender               0
customer.SeniorCitizen        0
customer.Partner              0
customer.Dependents           0
customer.tenure               0
phone.PhoneService            0
phone.MultipleLines           0
internet.InternetService      0
internet.OnlineSecurity       0
internet.OnlineBackup         0
internet.DeviceProtection     0
internet.TechSupport          0
internet.StreamingTV          0
internet.StreamingMovies      0
account.Contract              0
account.PaperlessBilling      0
account.PaymentMethod         0
account.Charges.Monthly       0
account.Charges.Total        11
dtype: int64


In [17]:
df = df.dropna(subset=['account.Charges.Total'])

In [18]:
print("Numero de Nulos \n", df.isnull().sum())

Numero de Nulos 
 customerID                   0
Churn                        0
customer.gender              0
customer.SeniorCitizen       0
customer.Partner             0
customer.Dependents          0
customer.tenure              0
phone.PhoneService           0
phone.MultipleLines          0
internet.InternetService     0
internet.OnlineSecurity      0
internet.OnlineBackup        0
internet.DeviceProtection    0
internet.TechSupport         0
internet.StreamingTV         0
internet.StreamingMovies     0
account.Contract             0
account.PaperlessBilling     0
account.PaymentMethod        0
account.Charges.Monthly      0
account.Charges.Total        0
dtype: int64


**4. Crear la columna de cuentas diarias (Opcional)**

In [19]:
df['cuentas_diarias'] = df['account.Charges.Monthly'] / 30
df

customerID Churn customer.gender  customer.SeniorCitizen  \
0     0002-ORFBO    No          Female                       0   
1     0003-MKNFE    No            Male                       0   
2     0004-TLHLJ   Yes            Male                       0   
3     0011-IGKFF   Yes            Male                       1   
4     0013-EXCHZ   Yes          Female                       1   
...          ...   ...             ...                     ...   
7262  9987-LUTYD    No          Female                       0   
7263  9992-RRAMN   Yes            Male                       0   
7264  9992-UJOEL    No            Male                       0   
7265  9993-LHIEB    No            Male                       0   
7266  9995-HOTOH    No            Male                       0   

     customer.Partner customer.Dependents  customer.tenure phone.PhoneService  \
0                 Yes                 Yes                9                Yes   
1                  No                  No                9                Yes   
2                  No                  No                4                Yes   
3                 Yes                  No               13                Yes   
4                 Yes                  No                3                Yes   
...               ...                 ...              ...                ...   
7262               No                  No               13                Yes   
7263              Yes                  No               22                Yes   
7264               No                  No                2                Yes   
7265              Yes                 Yes               67                Yes   
7266              Yes                 Yes               63                 No   

     phone.MultipleLines internet.InternetService  ...  \
0                     No                      DSL  ...   
1                    Yes                      DSL  ...   
2                     No              Fiber optic  ...   
3                     No              Fiber optic  ...   
4                     No              Fiber optic  ...   
...                  ...                      ...  ...   
7262                  No                      DSL  ...   
7263                 Yes              Fiber optic  ...   
7264                  No                      DSL  ...   
7265                  No                      DSL  ...   
7266    No phone service                      DSL  ...   

     internet.DeviceProtection internet.TechSupport internet.StreamingTV  \
0                           No                  Yes                  Yes   
1                           No                   No                   No   
2                          Yes                   No                   No   
3                          Yes                   No                  Yes   
4                           No                  Yes                  Yes   
...                        ...                  ...                  ...   
7262                        No                  Yes                   No   
7263                        No                   No                   No   
7264                        No                   No                   No   
7265                       Yes                  Yes                   No   
7266                       Yes                   No                  Yes   

     internet.StreamingMovies account.Contract account.PaperlessBilling  \
0                          No         One year                      Yes   
1                         Yes   Month-to-month                       No   
2                          No   Month-to-month                      Yes   
3                         Yes   Month-to-month                      Yes   
4                          No   Month-to-month                      Yes   
...                       ...              ...                      ...   
7262                       No         One year                       No   
7263                      Yes 

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7032 entries, 0 to 7266
Data columns (total 22 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   customerID                 7032 non-null   object 
 1   Churn                      7032 non-null   object 
 2   customer.gender            7032 non-null   object 
 3   customer.SeniorCitizen     7032 non-null   int64  
 4   customer.Partner           7032 non-null   object 
 5   customer.Dependents        7032 non-null   object 
 6   customer.tenure            7032 non-null   int64  
 7   phone.PhoneService         7032 non-null   object 
 8   phone.MultipleLines        7032 non-null   object 
 9   internet.InternetService   7032 non-null   object 
 10  internet.OnlineSecurity    7032 non-null   object 
 11  internet.OnlineBackup      7032 non-null   object 
 12  internet.DeviceProtection  7032 non-null   object 
 13  internet.TechSupport       7032 non-null   object 
 1

In [21]:
df.shape

(7032, 22)

**Se Agrega un nuevo atributo en el DATASET, "cuentas_diarias" para el diccionario de datos**

22. `'cuentas_diarias'`: Valor cargo en cuenta por día.

**El DATASET queda ahora con 7.032 filas y 22 columnas-**


 **5. Traducir las columnas y/o datos.**

In [22]:
#Ajustamos nombre de atributos
columnas = {      # Diccionario con los nuevos nombres de las columnas.
    'customerID': 'id',
    'customer.gender': 'genero',
    'customer.SeniorCitizen': 'tiene +60',
    'customer.Partner': 'posee_pareja',
    'customer.Dependents': 'posee_dependientes',
    'customer.tenure': 'tiempo_contrato',
    'phone.PhoneService': 'servicio_telefono',
    'phone.MultipleLines': 'multiples_lineas',
    'internet.InternetService': 'tipo_internet',
    'internet.OnlineSecurity': 'seguridad_online',
    'internet.OnlineBackup': 'backup_online',
    'internet.DeviceProtection': 'proteccion_dispositivo',
    'internet.TechSupport': 'soporte_tecnico',
    'internet.StreamingTV': 'streaming_tv',
    'internet.StreamingMovies': 'streaming_peliculas',
    'account.Contract': 'tipo_contrato',
    'account.PaperlessBilling': 'factura_digital',
    'account.PaymentMethod': 'metodo_pago',
    'account.Charges.Monthly': 'valor_mensal',
    'account.Charges.Total': 'total_cobrado'
}
df = df.rename(columns= columnas)
df

id Churn  genero  tiene +60 posee_pareja posee_dependientes  \
0     0002-ORFBO    No  Female          0          Yes                Yes   
1     0003-MKNFE    No    Male          0           No                 No   
2     0004-TLHLJ   Yes    Male          0           No                 No   
3     0011-IGKFF   Yes    Male          1          Yes                 No   
4     0013-EXCHZ   Yes  Female          1          Yes                 No   
...          ...   ...     ...        ...          ...                ...   
7262  9987-LUTYD    No  Female          0           No                 No   
7263  9992-RRAMN   Yes    Male          0          Yes                 No   
7264  9992-UJOEL    No    Male          0           No                 No   
7265  9993-LHIEB    No    Male          0          Yes                Yes   
7266  9995-HOTOH    No    Male          0          Yes                Yes   

      tiempo_contrato servicio_telefono  multiples_lineas tipo_internet  ...  \
0                   9               Yes                No           DSL  ...   
1                   9               Yes               Yes           DSL  ...   
2                   4               Yes                No   Fiber optic  ...   
3                  13               Yes                No   Fiber optic  ...   
4                   3               Yes                No   Fiber optic  ...   
...               ...               ...               ...           ...  ...   
7262               13               Yes                No           DSL  ...   
7263               22               Yes               Yes   Fiber optic  ...   
7264                2               Yes                No           DSL  ...   
7265               67               Yes                No           DSL  ...   
7266               63                No  No phone service           DSL  ...   

     proteccion_dispositivo soporte_tecnico streaming_tv streaming_peliculas  \
0                        No             Yes          Yes                  No   
1                        No              No           No                 Yes   
2                       Yes              No           No                  No   
3                       Yes              No          Yes                 Yes   
4                        No             Yes          Yes                  No   
...                     ...             ...          ...                 ...   
7262                     No             Yes           No                  No   
7263                     No              No           No                 Yes   
7264                     No              No           No                  No   
7265                    Yes             Yes           No                 Yes   
7266                    Yes              No          Yes                 Yes   

       tipo_contrato factura_digital       metodo_pago valor_mensal  \
0           One year             Yes      Mailed check        65.60   
1     Month-to-month              No      Mailed check        59.90   
2     Month-to-month             Yes  Electronic check        73.90   
3     Month-to-month             Yes  Electronic check        98.00   
4     Month-to-month             Yes      Mailed check        83.90   
...              ...             ...               ...          ...   
7262        One year              No      Mailed check        55.15   
7263  Month-to-month             Yes  Electronic check        85.10   
7264  Month-to-month             Yes      Mailed check        50.30   
7265        Two year              No      Mailed check        67.85   
7266        Two year              No  Electronic check        59.00   

     total_cobrado  cuentas_diarias  
0           593.30         2.186667  
1           542.40         1.996667  
2           280.85         2.463333  
3          1237.85         3.266667  
4           267.40         2.796667  
...            ...              ...  
7262        742.90         1.838333  
7263      

In [23]:
#Ajustando contenidos de atributos
df['Churn'] = df['Churn'].replace({'No': 'No', 'Yes': 'Sí'})
df['genero'] = df['genero'].replace({'Female': 'Femenino', 'Male': 'Masculino'})
df['posee_pareja'] = df['posee_pareja'].replace({'Yes': 'Sí', 'No': 'No'})
df['posee_dependientes'] = df['posee_dependientes'].replace({'Yes': 'Sí', 'No': 'No'})
df['servicio_telefono'] = df['servicio_telefono'].replace({'Yes': 'Sí', 'No': 'No'})
df['multiples_lineas'] = df['multiples_lineas'].replace({'No': 'No', 'Yes': 'Sí', 'No phone service': 'Sin servicio de teléfono'})
df['tipo_internet'] = df['tipo_internet'].replace({'No': 'No'})
df['tipo_contrato'] = df['tipo_contrato'].replace({'One year': 'Anual', 'Month-to-month': 'Mensual', 'Two year': 'Bienal'})
df.head()

id Churn     genero  tiene +60 posee_pareja posee_dependientes  \
0  0002-ORFBO    No   Femenino          0           Sí                 Sí   
1  0003-MKNFE    No  Masculino          0           No                 No   
2  0004-TLHLJ    Sí  Masculino          0           No                 No   
3  0011-IGKFF    Sí  Masculino          1           Sí                 No   
4  0013-EXCHZ    Sí   Femenino          1           Sí                 No   

   tiempo_contrato servicio_telefono multiples_lineas tipo_internet  ...  \
0                9                Sí               No           DSL  ...   
1                9                Sí               Sí           DSL  ...   
2                4                Sí               No   Fiber optic  ...   
3               13                Sí               No   Fiber optic  ...   
4                3                Sí               No   Fiber optic  ...   

  proteccion_dispositivo soporte_tecnico streaming_tv streaming_peliculas  \
0                     No             Yes          Yes                  No   
1                     No              No           No                 Yes   
2                    Yes              No           No                  No   
3                    Yes              No          Yes                 Yes   
4                     No             Yes          Yes                  No   

  tipo_contrato factura_digital       metodo_pago valor_mensal total_cobrado  \
0         Anual             Yes      Mailed check         65.6        593.30   
1       Mensual              No      Mailed check         59.9        542.40   
2       Mensual             Yes  Electronic check         73.9        280.85   
3       Mensual             Yes  Electronic check         98.0       1237.85   
4       Mensual             Yes      Mailed check         83.9        267.40   

   cuentas_diarias  
0         2.186667  
1         1.996667  
2         2.463333  
3         3.266667  
4         2.796667  

[5 rows x 22 columns]

In [24]:
#Traduccion de columnas (atributos)
columnas_a_traducir = ['seguridad_online', 'backup_online', 'proteccion_dispositivo', 'soporte_tecnico', 'streaming_tv', 'streaming_peliculas']
mapeo = {'No': 'No', 'Yes': 'Sí', 'No internet service': 'Sin servicio de internet'}

for col in columnas_a_traducir:
    df[col] = df[col].replace(mapeo)

df.head()

id Churn     genero  tiene +60 posee_pareja posee_dependientes  \
0  0002-ORFBO    No   Femenino          0           Sí                 Sí   
1  0003-MKNFE    No  Masculino          0           No                 No   
2  0004-TLHLJ    Sí  Masculino          0           No                 No   
3  0011-IGKFF    Sí  Masculino          1           Sí                 No   
4  0013-EXCHZ    Sí   Femenino          1           Sí                 No   

   tiempo_contrato servicio_telefono multiples_lineas tipo_internet  ...  \
0                9                Sí               No           DSL  ...   
1                9                Sí               Sí           DSL  ...   
2                4                Sí               No   Fiber optic  ...   
3               13                Sí               No   Fiber optic  ...   
4                3                Sí               No   Fiber optic  ...   

  proteccion_dispositivo soporte_tecnico streaming_tv streaming_peliculas  \
0                     No              Sí           Sí                  No   
1                     No              No           No                  Sí   
2                     Sí              No           No                  No   
3                     Sí              No           Sí                  Sí   
4                     No              Sí           Sí                  No   

  tipo_contrato factura_digital       metodo_pago valor_mensal total_cobrado  \
0         Anual             Yes      Mailed check         65.6        593.30   
1       Mensual              No      Mailed check         59.9        542.40   
2       Mensual             Yes  Electronic check         73.9        280.85   
3       Mensual             Yes  Electronic check         98.0       1237.85   
4       Mensual             Yes      Mailed check         83.9        267.40   

   cuentas_diarias  
0         2.186667  
1         1.996667  
2         2.463333  
3         3.266667  
4         2.796667  

[5 rows x 22 columns]

In [27]:
#Traducción del registros para columna "metodo_pago"
df['metodo_pago'] = df['metodo_pago'].replace({
    'Mailed check': 'Cheque enviado por correo',
    'Electronic check': 'Cheque electrónico',
    'Credit card (automatic)': 'Tarjeta de crédito (automático)',
    'Bank transfer (automatic)': 'Transferencia bancaria (automática)'
})
df.head()

id Churn     genero  tiene +60 posee_pareja posee_dependientes  \
0  0002-ORFBO    No   Femenino          0           Sí                 Sí   
1  0003-MKNFE    No  Masculino          0           No                 No   
2  0004-TLHLJ    Sí  Masculino          0           No                 No   
3  0011-IGKFF    Sí  Masculino          1           Sí                 No   
4  0013-EXCHZ    Sí   Femenino          1           Sí                 No   

   tiempo_contrato servicio_telefono multiples_lineas tipo_internet  ...  \
0                9                Sí               No           DSL  ...   
1                9                Sí               Sí           DSL  ...   
2                4                Sí               No   Fiber optic  ...   
3               13                Sí               No   Fiber optic  ...   
4                3                Sí               No   Fiber optic  ...   

  proteccion_dispositivo soporte_tecnico streaming_tv streaming_peliculas  \
0                     No              Sí           Sí                  No   
1                     No              No           No                  Sí   
2                     Sí              No           No                  No   
3                     Sí              No           Sí                  Sí   
4                     No              Sí           Sí                  No   

  tipo_contrato factura_digital                metodo_pago valor_mensal  \
0         Anual             Yes  Cheque enviado por correo         65.6   
1       Mensual              No  Cheque enviado por correo         59.9   
2       Mensual             Yes         Cheque electrónico         73.9   
3       Mensual             Yes         Cheque electrónico         98.0   
4       Mensual             Yes  Cheque enviado por correo         83.9   

  total_cobrado  cuentas_diarias  
0        593.30         2.186667  
1        542.40         1.996667  
2        280.85         2.463333  
3       1237.85         3.266667  
4        267.40         2.796667  

[5 rows x 22 columns]

In [29]:
df

id Churn     genero  tiene +60 posee_pareja posee_dependientes  \
0     0002-ORFBO    No   Femenino          0           Sí                 Sí   
1     0003-MKNFE    No  Masculino          0           No                 No   
2     0004-TLHLJ    Sí  Masculino          0           No                 No   
3     0011-IGKFF    Sí  Masculino          1           Sí                 No   
4     0013-EXCHZ    Sí   Femenino          1           Sí                 No   
...          ...   ...        ...        ...          ...                ...   
7262  9987-LUTYD    No   Femenino          0           No                 No   
7263  9992-RRAMN    Sí  Masculino          0           Sí                 No   
7264  9992-UJOEL    No  Masculino          0           No                 No   
7265  9993-LHIEB    No  Masculino          0           Sí                 Sí   
7266  9995-HOTOH    No  Masculino          0           Sí                 Sí   

      tiempo_contrato servicio_telefono          multiples_lineas  \
0                   9                Sí                        No   
1                   9                Sí                        Sí   
2                   4                Sí                        No   
3                  13                Sí                        No   
4                   3                Sí                        No   
...               ...               ...                       ...   
7262               13                Sí                        No   
7263               22                Sí                        Sí   
7264                2                Sí                        No   
7265               67                Sí                        No   
7266               63                No  Sin servicio de teléfono   

     tipo_internet  ... proteccion_dispositivo soporte_tecnico streaming_tv  \
0              DSL  ...                     No              Sí           Sí   
1              DSL  ...                     No              No           No   
2      Fiber optic  ...                     Sí              No           No   
3      Fiber optic  ...                     Sí              No           Sí   
4      Fiber optic  ...                     No              Sí           Sí   
...            ...  ...                    ...             ...          ...   
7262           DSL  ...                     No              Sí           No   
7263   Fiber optic  ...                     No              No           No   
7264           DSL  ...                     No              No           No   
7265           DSL  ...                     Sí              Sí           No   
7266           DSL  ...                     Sí              No           Sí   

     streaming_peliculas tipo_contrato factura_digital  \
0                     No         Anual             Yes   
1                     Sí       Mensual              No   
2                     No       Mensual             Yes   
3                     Sí       Mensual             Yes   
4                     No       Mensual             Yes   
...                  ...           ...             ...   
7262                  No         Anual              No   
7263                  Sí       Mensual             Yes   
7264                  No       Mensual             Yes   
7265                  Sí        Bienal              No   
7266                  Sí        Bienal              No   

                    metodo_pago valor_mensal total_cobrado  cuentas_diarias  
0     Cheque enviado por correo        65.60        593.30         2.186667  
1     Cheque enviado por correo        59.90        542.40         1.996667  
2            Cheque electrónico        73.90        280.85         2.463333  
3            Cheque electrónico        98.00       1237.85         3.266667  
4     Cheque enviado por correo        83.90        267.40         2.796667  
...                         ...          ...           ...              ...  
7262  Cheque enviado por correo 

In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7032 entries, 0 to 7266
Data columns (total 22 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   id                      7032 non-null   object 
 1   Churn                   7032 non-null   object 
 2   genero                  7032 non-null   object 
 3   tiene +60               7032 non-null   int64  
 4   posee_pareja            7032 non-null   object 
 5   posee_dependientes      7032 non-null   object 
 6   tiempo_contrato         7032 non-null   int64  
 7   servicio_telefono       7032 non-null   object 
 8   multiples_lineas        7032 non-null   object 
 9   tipo_internet           7032 non-null   object 
 10  seguridad_online        7032 non-null   object 
 11  backup_online           7032 non-null   object 
 12  proteccion_dispositivo  7032 non-null   object 
 13  soporte_tecnico         7032 non-null   object 
 14  streaming_tv            7032 non-null   objec

In [31]:
df.shape

(7032, 22)

**🔍** **Resultados de nuevo dataset ajustado **

  - **filas** - `'7.031'`: Cantidad de registros.
  - **columnas** - `'22'`: Atributos de los datos.

_____________________________________________________________________________

Se ajustan los datos para asegurar que estén completos y coherentes, preparándolos para las siguientes etapas del análisis, donde se ha generado un nuevo dataset con el siguiente diccionario de datos:

**DICCIONARIO DE DATOS -Dataset ajustado con nuevos nombre de atributos traducidos**

1. `'id'`: Identificador único del cliente.
2. `'Churn'`: Indica si el cliente canceló el servicio (`Sí` o `No`).
3. `'genero'`: Género del cliente (`Male` o `Female`).
4. `'tiene +60'`: Indica si el cliente es una persona mayor (0 = No, 1 = Sí).
5. `'posee_pareja'`: Si el cliente tiene pareja (`Sí` o `No`).
6. `'posee_dependientes'`: Si el cliente tiene dependientes (`Sí` o `No`).
7. `'tiempo_contrato'`: Tiempo de permanencia como cliente (en meses).
8. `'servicio_telefono'`: Indica si el cliente posee servicio de teléfono (`Yes` o `No`, `Sin servicio de teléfono`).
9. `'multiples_lineas'`: Si posee múltiples líneas telefónicas. (`Sí` o `No`,  `Sin servicio de teléfono`).
10. `'tipo_internet'`: Servicio de internet (`DSL`, `Fiber optic`, `No`).
11. `'seguridad_online'`: Servicio internet de seguridad (`No`, `Sí`, `Sin servicio de internet'`).
12. `'backup_online '`: Servicio internet de respaldo (`No`, `Sí`, `Sin servicio de internet`).
13. `'proteccion_dispositivo'`: Servicio internet proteciòn (`No`, `Sí`, `Sin servicio de internet`).
14. `'soporte_tecnico'`: Servicio internet soporte tècnico (`No`, `Sí`, `Sin servicio de internet`).
15. `'streaming_tv'`: Servicio internet streaming de televisión (`No`, `Sí`, `Sin servicio de internet`).
16. `'streaming_peliculas'`: Servicio internet streaming de peliculas (`No`, `Yes`, `Sin servicio de internet`).
17. `'tipo_contrato'`: Tipo de contrato (`Mensual`, `Anual`, `Bianual`).
18. `'factura_digital'`: Si el cliente recibe facturas electrónicas (`Sí` o `No`).
19. `'metodo_pago'`: Método de pago (`Transferencia bancaria (automática)`, `Tarjeta de crédito (automático)`, `Cheque enviado por correo`, `Cheque electrónico`).
20. `'valor_mensual'`: Valor mensual cobrado.
21. `'total_cobrado'`: Valor total pagado por el cliente.
22. `'cuentas_diarias'`: Valor cargo en cuenta por día.

_____________________________________________________________________________

** ------- Actividades Tranformación FINALIZADA --- **

_____________________________________________________________________________





#📊 Carga y análisis

#📄Informe final